### ISSUE TITLE
Nginx Modul support

### ISSUE THREAD

**Issue by wuast94** [NginxProxyManager/nginx-proxy-manager#78]
Would be nice to have the possibility to activate nginx modules. 

For example:
http_geoip_module
Or
http_stub_status_module

**Comment by jc21** [comment]
Geoip is already planned to be added in #46. Any future modules might have to be included as part of docker custom builds, as in the non-x86_64 builds, nginx is built from source.

**Comment by wuast94** [comment]
after a long night i have new infos :D 

thoose modules are alred activ but it was quit hard to get it runnig. 

1. you need to edit some files that are not mapped in /config of container but this isnt  big problem:

/etc/nginx/conf.d/status.conf (to enable the /nginx_status page)
/etc/nginx/nginx.conf (for geoip and i do some custom logging for more infos) 
/etc/nginx/geoip (to dd the geoip files .. i tell more bout it in #46 )

i think it would be nice to symlink /etc/nginx/ to /config/nginx/config (or something like that) 

2. when i mapped everthing i must add 
`access_log /config/log/nginx/accesscust.log custom;` 
for custom logging in every proxy host i have. 
and as far as i understand it is a thing i cant put in advanced setting if i edit a proxy. or can i ? 

i think i can made a tutorial or something like that for my problem. in main its used for get data from nginx .. and visualize it with grafana :)

**Comment by jc21** [comment]
To simplify a possible generic solution here, you want to be able to add your own configuration items to the main `nginx` block, not inside the `server` block?

Specific to the status page plugin, it might be possible to make that information available from within the applicaiton interface.

**Comment by wuast94** [comment]
Yes..would be nice to be able to edit the nginx block as you told it :) 

I think too yes. I get this working and it's nice to see what happens over nginx. The Only Stat im missing ist the bandwidth (up and down stream) but i don't think this is possible over ngix..but you can get this information over docker. But only for the container not for services/Proxys inside the container. 

And yes..a interface in the container could be nice to..I'm using telegraf/influxdb/grafana for visualizing Infos from nginx and other services :)

**Comment by jc21** [comment]
Yeah I don't think you're going to get specific hostname traffic statistics without an logfile analyzer, ELK or prometheus etc. I've thought about having another docker image that accompanies this project to provide this, but I haven't had any time to look in to it.

**Comment by pitsch** [comment]
whitelisting firewall to prevent XSL/SQL attacks, has a learning mode when you add a new service. 
https://github.com/nbs-system/naxsi

**Comment by OhHeyAlan** [comment]
I’ve tried to recreate this app (the base image) based on Nginx/1.15.12-Alpine and complied with OpenSSL 1.1.1c and the build is successful but I’m not sure how to get it to work with the final image... 

Would love to harden the security of this container and make it easier to add new features/customize.

**Comment by jc21** [comment]
As a proof of concept, you can fork this repo, modify the `Dockerfile` to use your new base image, create a PR and my CI will create a working docker image for you to test.

**Comment by OhHeyAlan** [comment]
I’m under the weather right now but I’ll give that a try. I use GitLab-CI and I’ve been able to build a few new images and they run but without web UI. 

Lately the build fails once it gets to the ‘ADD dist        /app/dist’ step. Says the directory doesn’t exist.

**Comment by ansuz07** [comment]
@wuast94 - Could you post a short guide on how you got the status/geoip working?  I'm having a beast of a time with it.

Thanks!

**Comment by wuast94** [comment]
> @wuast94 - Could you post a short guide on how you got the status/geoip working? I'm having a beast of a time with it.
> 
> Thanks!

Here I have written how I get it working for me. Can you tell me how far you come and where your problems are? 
https://github.com/jc21/nginx-proxy-manager/issues/78#issuecomment-465956712

**Comment by ansuz07** [comment]
@wuast94 Thanks for the reply.  I've managed to get the status page working, but I'm stuck on the GeoIP bit.

I've been trying to follow this guide:  https://medium.com/faun/total-nginx-monitoring-with-application-performance-and-a-bit-more-using-8fc6d731051b but when I add the geoip references to the nginx.conf file and run nginx -t I get the following:

     nginx: [emerg] unknown directive "geoip_country" in /etc/nginx/nginx.conf:42

What did you add to your nginx.conf file to get this working?  Thanks!

**Comment by tuxmeister** [comment]
> after a long night i have new infos :D
> 
> thoose modules are alred activ but it was quit hard to get it runnig.
> 
> 1. you need to edit some files that are not mapped in /config of container but this isnt  big problem:
> 
> /etc/nginx/conf.d/status.conf (to enable the /nginx_status page)
> /etc/nginx/nginx.conf (for geoip and i do some custom logging for more infos)
> /etc/nginx/geoip (to dd the geoip files .. i tell more bout it in #46 )
> 
> i think it would be nice to symlink /etc/nginx/ to /config/nginx/config (or something like that)
> 
> 1. when i mapped everthing i must add
>    `access_log /config/log/nginx/accesscust.log custom;`
>    for custom logging in every proxy host i have.
>    and as far as i understand it is a thing i cant put in advanced setting if i edit a proxy. or can i ?
> 
> i think i can made a tutorial or something like that for my problem. in main its used for get data from nginx .. and visualize it with grafana :)

Did you manage to put together a tutorial for this? I am also interested in extracting logs through telegraf for a TIG setup. I'm mostly interested in your custom logging solution for all the proxy hosts into a single file. Cheers!

**Comment by wuast94** [comment]
@tuxmeister everything relevant is here and in #46 

I'm not using this setup anymore..or more specific the geoip module. This was more becouse of traffic logging per sub domain wich isn't possible in the free or community version of nginx. Could be implemented in nginxproxymanager but I think that's a huge amount of work.
